perf/collector_libpfm.go

```diff
@@ -0,0 +1,250 @@
+	for name, files := range c.cpuFiles {
+		for cpu, file := range files {
+			buf := make([]byte, 32)
+			_, err := file.Read(buf)
+			if err != nil {
+				klog.Warningf("Unable to read from perf_event file (event: %q, CPU: %d) for %q", name, cpu, c.cgroupPath)
+				continue
+			}
+			perfData := &ReadFormat{}
+			reader := bytes.NewReader(buf)
+			err = binary.Read(reader, binary.LittleEndian, perfData)
+			if err != nil {
+				klog.Warningf("Unable to decode from binary format read from perf_event file (event: %q, CPU: %d) for %q", name, cpu, c.cgroupPath)
+			}
+			klog.V(4).Infof("Read metric for event %q for cpu %d from cgroup %q: %d", name, cpu, c.cgroupPath, perfData.Value)
+			scalingRatio := 1.0
+			if perfData.TimeEnabled != 0 {
+				scalingRatio = float64(perfData.TimeRunning) / float64(perfData.TimeEnabled)
+			}
+			stat := info.PerfStat{
+				Value:        uint64(float64(perfData.Value) / scalingRatio),
+				Name:         name,
+				ScalingRatio: scalingRatio,
+				Cpu:          cpu,
+			}
+			stats.PerfStats = append(stats.PerfStats, stat)
+		}
+	}
+	return nil
+}
+
+func (c *collector) Setup() {
+	cgroup, err := os.Open(c.cgroupPath)
+	if err != nil {
+		klog.Warningf("Unable to open cgroup directory %s: %s", c.cgroupPath, err)
+	}
+	defer cgroup.Close()
+
+	c.cpuFilesLock.Lock()
+	defer c.cpuFilesLock.Unlock()
+	cgroupFd := int(cgroup.Fd())
+	for _, group := range c.events.Events {
+		if len(group) == 1 {
+			customEvent, ok := c.events.eventToCustomEvent[group[0]]
+			if ok {
+				c.setupRawNonGrouped(customEvent, cgroupFd)
+			} else {
+				c.setupNonGrouped(string(group[0]), cgroupFd)
+			}
+		} else {
+			klog.Info("Grouped events are not supported yet")
+		}
+	}
+}
+
+func (c *collector) setupRawNonGrouped(event *CustomEvent, cgroup int) {
+	klog.V(4).Infof("Setting up non-grouped raw perf event %#v", event)
+	config := createPerfEventAttr(*event)
+	c.registerEvent(config, string(event.Name), cgroup)
+}
+
+func (c *collector) registerEvent(config *unix.PerfEventAttr, name string, cgroup int) {
+	var cpu int
+	for cpu = 0; cpu < c.numCores; cpu++ {
+		pid, groupFd, flags := cgroup, -1, unix.PERF_FLAG_FD_CLOEXEC|unix.PERF_FLAG_PID_CGROUP
+		fd, err := unix.PerfEventOpen(config, pid, cpu, groupFd, flags)
+		if err != nil {
+			klog.Errorf("Setting up perf event %#v failed: %q", config, err)
+			return
+		}
+		perfFile := os.NewFile(uintptr(fd), name)
+		if perfFile == nil {
+			klog.Warningf("Unable to create os.File from file descriptor %#v", fd)
+		}
+
+		c.addEventFile(name, cpu, perfFile)
+	}
+}
+
+func (c *collector) addEventFile(name string, cpu int, perfFile *os.File) {
+	_, ok := c.cpuFiles[name]
+	if !ok {
+		c.cpuFiles[name] = map[int]readerCloser{}
+	}
+	c.cpuFiles[name][cpu] = perfFile
+}
+
+func (c *collector) setupNonGrouped(name string, cgroup int) error {
+	if !isLibpfmInitialized {
+		klog.Warning("libpfm4 is not initialized, cannot proceed with setting perf events up")
```

Comment:
We should probably handle the error in the caller and log there.  That way we can provide additional context about the call.

Same goes for other places where we log and return the same error.